File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

public class Main {

    
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        while (t-- > 0) {
            int[] dp = new int[12];
            int n = Integer.parseInt(br.readLine());
            dp[0] = 1;
            for (int i = 1; i <= n; i++) {
                dp[i] += dp[i - 1];
                if (i - 2 >= 0) dp[i] += dp[i - 2];
                if (i - 3 >= 0) dp[i] += dp[i - 3];
            }
            sb.append(dp[n]).append("\n");
        }
        System.out.print(sb);
    }
}

File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.kt
/**
 * 순서를 고려해서 경우의 수를 구함 -> 과거의 결과가 미래의 구성에 간섭이 없음
 */

fun main() = with(System.`in`.bufferedReader()) {
    val d = IntArray(11)
    d[0] = 1
    for (i in 1..10) {
        d[i] += (d[i - 1])
        if (i - 2 >= 0) d[i] += d[i - 2]
        if (i - 3 >= 0) d[i] += d[i - 3]
    }

    print(IntArray(readLine().toInt()) {
        d[readLine().toInt()]
    }.joinToString("\n"))
}